documents/repository.go

```diff
@@ -1,39 +1,195 @@
+type LegacyRepository interface {
 	// GetKey will prepare the the identifier key from ID
 	GetKey(id []byte) (key []byte)
 
 	// GetByID finds the doc with identifier and marshals it into message
 	LoadByID(id []byte, model Model) error
-}
 
-// Checker interface can be implemented by any type that handles if document exists
-type Checker interface {
 	// Exists checks for document existence
 	// True if exists else false
 	Exists(id []byte) bool
-}
 
-// Creator interface can be implemented by any type that handles document creation
-type Creator interface {
 	// Create stores the initial document
 	// If document exist, it errors out
 	Create(id []byte, model Model) error
-}
 
-// Updater interface can be implemented by any type that handles document update
-type Updater interface {
 	// Update updates the already stored document
 	// errors out when document is missing
 	Update(id []byte, model Model) error
 }
 
-// Repository should be implemented by any type that wants to store a document in key-value storage
+// Repository defines the required methods for a document repository.
+// Can be implemented by any type that stores the documents. Ex: levelDB, sql etc...
 type Repository interface {
-	Checker
-	Loader
-	Creator
-	Updater
+	// Exists checks if the id, owned by tenantID, exists in DB
+	Exists(tenantID, id []byte) bool
+
+	// Get returns the Model associated with ID, owned by tenantID
+	Get(tenantID, id []byte) (Model, error)
+
+	// Create creates the model if not present in the DB.
+	// should error out if the document exists.
+	Create(tenantID, id []byte, model Model) error
+
+	// Update strictly updates the model.
+	// Will error out when the model doesn't exist in the DB.
+	Update(tenantID, id []byte, model Model) error
+
+	// Register registers the model so that the DB can return the document without knowing the type
+	Register(model Model)
+}
+
+// levelDBRepo implements Repository using LevelDB as storage layer
+type levelDBRepo struct {
+	db     *leveldb.DB
+	models map[string]reflect.Type
+	mu     sync.RWMutex // to protect the models
+}
+
+// value is an internal representation of how levelDb stores the model.
+type value struct {
+	Type string          `json:"type"`
+	Data json.RawMessage `json:"data"`
+}
+
+// NewLevelDBRepository returns levelDb implementation of Repository
+func NewLevelDBRepository(db *leveldb.DB) Repository {
+	return &levelDBRepo{
+		db:     db,
+		models: make(map[string]reflect.Type),
+	}
+}
+
+// getKey returns tenantID+id
+func getKey(tenantID, id []byte) []byte {
+	return append(tenantID, id...)
+}
+
+// Exists returns true if the id, owned by tenantID, exists.
+func (l *levelDBRepo) Exists(tenantID, id []byte) bool {
+	key := getKey(tenantID, id)
+	res, err := l.db.Has(key, nil)
+	// TODO check this
+	if err != nil {
+		return false
+	}
+
+	return res
+}
+
+// getModel returns a new instance of the type mt.
+func (l *levelDBRepo) getModel(mt string) (Model, error) {
```

Comment:
maybe:
```suggestion
func (l *levelDBRepo) getModel(modelType string) (Model, error) {
```

Reply:
Shorter variables are better since this is an internal function IMO